threedi_models_and_simulations/widgets/log_in.py

```diff
@@ -43,6 +51,143 @@ def wrapper(self):
     return wrapper
 
 
+class AuthorizationCodeInterceptor(QObject):
+    """Authorization code interceptor class."""
+
+    code_intercepted = pyqtSignal(str)
+    code_interception_failed = pyqtSignal(Exception)
+
+    def __init__(self, authorization_handler, state=None):
+        super().__init__()
+        self.authorization_handler = authorization_handler
+        self.state = state
+        self.tcp_server = None
+
+    def start_listening(self, host=QHostAddress.LocalHost, port=1234):
+        """Start listening for an incoming connections."""
+        self.tcp_server = QTcpServer(self)
+        if not self.tcp_server.listen(host, port):
+            raise RuntimeError(f"Can't listen port: '{port}'!")
+        self.tcp_server.newConnection.connect(self.intercept_code)
+
+    def stop_listening(self):
+        """Stop listening for an incoming connections."""
+        if self.tcp_server is not None and self.tcp_server.isListening():
+            self.tcp_server.close()
+
+    def intercept_code(self):
+        """Intercept an authorization code from the incoming connection."""
+        try:
+            self.tcp_server.newConnection.disconnect(self.intercept_code)
+            tcp_socket = self.tcp_server.nextPendingConnection()
+            tcp_socket.waitForReadyRead()
+            incoming_data = tcp_socket.readAll().data()
+            incoming_data_str = str(incoming_data, "utf-8")
+            params_search_results = re.search(r"code=(?P<code>[^&]+)&state=(?P<state>\S+)", incoming_data_str)
+            params = params_search_results.groupdict()
+            authorization_code = params["code"]
+            state = params["state"]
+            if self.state:
+                assert state == self.state
+            tcp_socket.disconnected.connect(tcp_socket.deleteLater)
+            tcp_socket.disconnectFromHost()
+            self.code_intercepted.emit(authorization_code)
+        except Exception as error:
+            self.code_interception_failed.emit(error)
+        finally:
+            self.stop_listening()
+
+
+class AuthorizationHandler(QObject):
+    """Class for handling OAuth2 + PKCE authorization."""
+
+    tokens_acquired = pyqtSignal(str, str)
+
+    AUTHORIZATION_ENDPOINT = "https://auth.lizard.net/oauth2/authorize"
+    TOKEN_ENDPOINT = "https://auth.lizard.net/oauth2/token"
+    REDIRECT_URI = "http://localhost:1234/"
+    CLIENT_ID = "73d21iv9pu333mjqpbguipa7pi"
+    SCOPE = "staging.3di.live/*:readwrite"
+    STATE = "fooobarbaz"
```

Comment:
State must be randomly generated at the moment you do the request.

Something like

```
>>> import secrets
>>> state = secrets.token_urlsafe(22) 
'8d9bad5b43259c6ee27d9aadc7b832'
```